### PR TITLE
CUDNN v4 upgrade.  Optionally, CNMEM used for GPU pool allocation.

### PR DESCRIPTION
**Pull request by borisfom**
Note: CNMEM(https://github.com/NVIDIA/cnmem.git) submodule added at caffe/cnmem.

Attention:

1. ./cnmem is not getting populated by default with 'git clone' or pull.
   To populate added CNMEM submodule, do 'git submodule update'.
   Alernatively, CNMEM can be pulled externally from https://github.com/NVIDIA/cnmem.git.
   Caffe does not make any assumptions about actual CNMEM code location at the time.

2. CNMEM has to be built separately. Refer to CNMEM's README.md

3. CMNEM pool is turned on by default. It can be turned off by adding 'false' parameter to the constructor of caffe::MemoryManager object  in caffe_main.cpp and/or test_caffe_main.cpp.

### CODE REVIEW

**Comment by lukeyeager** [comment]
I built it with cuDNN v3 with no errors, and it passed all Caffe and DIGITS tests. So that looks good so far.

**Comment by lukeyeager** [comment]
Looks good with v4 as well. No issues so far when testing multi-GPU, HDF5, LeNet, AlexNet, GoogLeNet, etc.

As an added bonus #16, #23, #28 and #38 seems to all be fixed too!

**Comment by lukeyeager** [comment]
Check your TravisCI build errors. Some of them seem to be simple lint errors. Try `make lint`.

**Comment by borisfom** [comment]
Yay!

File: .travis.yml [hunk 1]
```diff
@@ -8,6 +8,9 @@ env:
     - WITH_CUDA=true WITH_CMAKE=true WITH_IO=true
     - WITH_CUDA=false WITH_CMAKE=false WITH_IO=false
     - WITH_CUDA=false WITH_CMAKE=true WITH_IO=false PYTHON_VERSION=3
+#  Currently there is no way to install cudnn via apt-get. Uncomment wjen it's available.
```
Comment (lukeyeager) [inline .travis.yml]: check spelling

File: Makefile [hunk 2]
```diff
@@ -241,12 +259,12 @@ endif
 ifeq ($(LINUX), 1)
 	CXX ?= /usr/bin/g++
 	GCCVERSION := $(shell $(CXX) -dumpversion | cut -f1,2 -d.)
-	# older versions of gcc are too dumb to build boost with -Wuninitalized
+        # older versions of gcc are too dumb to build boost with -Wuninitalized
```
Comment (lukeyeager) [inline Makefile]: Lots of unnecessary whitespace changes in this file

File: README.md [hunk 3]
```diff
@@ -1,5 +1,8 @@
 # Caffe
 
+[![Build Status](https://travis-ci.org/BVLC/caffe.svg?branch=master)](https://travis-ci.org/BVLC/caffe)
+[![License](https://img.shields.io/badge/license-BSD-blue.svg)](LICENSE)
+
```
Comment (lukeyeager) [inline README.md]: Not related to this PR

File: docs/tutorial/layers.md [hunk 4]
```diff
@@ -39,7 +39,7 @@ In contrast, other layers (with few exceptions) ignore the spatial structure of
     - `n * c_i * h_i * w_i`
 * Output
     - `n * c_o * h_o * w_o`, where `h_o = (h_i + 2 * pad_h - kernel_h) / stride_h + 1` and `w_o` likewise.
-* Sample (as seen in `./examples/imagenet/imagenet_train_val.prototxt`)
+* Sample (as seen in `./models/bvlc_reference_caffenet/train_val.prototxt`)
```
Comment (lukeyeager) [inline docs/tutorial/layers.md]: Not related to this PR

File: src/caffe/common.cpp [hunk 5]
```diff
@@ -192,7 +210,6 @@ void Caffe::DeviceQuery() {
   return;
 }
 
-
```
Comment (lukeyeager) [inline src/caffe/common.cpp]: Unnecessary whitespace change

File: matlab/+caffe/io.m [hunk 6]
```diff
@@ -29,5 +29,13 @@
...
       mean_data = caffe_('read_mean', mean_proto_file);
     end
+    function write_mean(mean_data, mean_proto_file)
+      % write_mean(mean_data, mean_proto_file)
+      %   write image mean data to binaryproto file
+      %   mean_data should be W x H x C with BGR channels
+      CHECK(ischar(mean_proto_file), 'mean_proto_file must be a string');
+      CHECK(isa(mean_data, 'single'), 'mean_data must be a SINGLE matrix');
+      caffe_('write_mean', mean_data, mean_proto_file);
+    end
```
Comment (lukeyeager) [inline matlab/+caffe/io.m]: Not related to this PR

File: matlab/+caffe/private/caffe_.cpp [hunk 7]
```diff
@@ -478,6 +478,29 @@ static void read_mean(MEX_ARGS) {
...
+  else
+    channels = 1;
+  Blob<float> data_mean(1, channels, height, width);
+  mx_mat_to_blob(prhs[0], &data_mean, DATA);
+  BlobProto blob_proto;
+  data_mean.ToProto(&blob_proto, false);
+  WriteProtoToBinaryFile(blob_proto, mean_proto_file);
+  mxFree(mean_proto_file);
+}
+
```
Comment (lukeyeager) [inline matlab/+caffe/private/caffe_.cpp]: Not related to this PR

**Comment by lukeyeager** [comment]
Woohoo, all the tests are passing! Now lets clean up the commit history and get this merged!

* There are several commits that are unrelated to this PR - how did those get in there?
* There are a lot of useless "fix" and "merge" commits - let's squash those together.

We don't need to get it down to a single commit - we can preserve authorship - but let's pare it down to a reasonable number of commits with clear commit messages explaining the contribution of each.

Also, the `.gitmodules` file is still there even after you removed it. Not sure what's going on there.

File: Makefile (hunk 2 above)
Comment (borisfom) [inline Makefile]: This is Makefile - tabs vs. spaces do matter :)

File: matlab/+caffe/io.m (hunk 6 above)
Comment (borisfom) [inline matlab/+caffe/io.m]: Let's keep this one - it's adding a function to the table, difficult to make sure no one calls it

File: matlab/+caffe/private/caffe_.cpp (hunk 7 above)
Comment (borisfom) [inline matlab/+caffe/private/caffe_.cpp]: Same as above

File: .travis.yml (hunk 1 above)
Comment (borisfom) [inline .travis.yml]: Fixed.

File: README.md (hunk 3 above)
Comment (borisfom) [inline README.md]: Done

File: docs/tutorial/layers.md (hunk 4 above)
Comment (borisfom) [inline docs/tutorial/layers.md]: Done

File: src/caffe/common.cpp (hunk 5 above)
Comment (borisfom) [inline src/caffe/common.cpp]: Done

File: matlab/+caffe/io.m (hunk 6 above)
Comment (lukeyeager) [inline matlab/+caffe/io.m]: I'm fine with including this and other upstream fixes in the `0.14` release, but I'd rather bring them in with a separate pull request. Let's stay focused with this PR.

**Comment by borisfom** [comment]
Done going through the notes - I have removed .gitmodules, too!
Please squash commits as you see fit!

**Comment by lukeyeager** [comment]
Awesome thanks. Ok, I'll give it a shot.

**Comment by lukeyeager** [comment]
Ok I give up. If I want to remove or squash commits I have to use `git rebase -i`, but I can't do that without removing all merge commits and attempting to turn the history into a linear stream of commits, which leads to lots of tricky merge commits. Not worth it.

On the bright side, we still have a working cuDNN v4 integration!

**Comment by borisfom** [comment]
Great news, thanks Luke!

File: Makefile [hunk 8]
```diff
@@ -7,12 +7,27 @@ $(error $(CONFIG_FILE) not found. See $(CONFIG_FILE).example.)
...
+endif
+
+ifeq ($(USE_CUDNN),1)
+# CNMEM is ON by default in CUDNN is ON
+  ifeq ($(USE_CNMEM),)
+    USE_CNMEM=1
+  endif
+endif
+
+PROJECT_DIR=$(PWD)
```
Comment (lukeyeager) [inline Makefile]: @borisfom why did you add this `PROJECT_DIR` variable? It broke my debian packaging. Is it really necessary?

Comment (borisfom) [inline Makefile]: I needed so set paths to absolute, let’s discuss this.
